storage-proofs/src/hybrid_merkle.rs

```diff
@@ -0,0 +1,1445 @@
+    alpha_tree: MerkleTree<AH::Domain, AH::Function>,
+    root: AH::Domain,
+    n_layers: usize,
+    height: usize,
+    alpha_tree_height: usize,
+    beta_tree_height: usize,
+    n_leaves: usize,
+    n_leaves_per_beta_tree: usize,
+    n_beta_trees: usize,
+    n_nodes: usize,
+    last_node_index_per_layer: Vec<usize>,
+    _ah: PhantomData<AH>,
+    _bh: PhantomData<BH>,
+}
+
+impl<AH, BH> Debug for HybridMerkleTree<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let beta_trees_debug: Vec<MerkleTreeRootDebug<BH::Domain>> = self
+            .beta_trees
+            .iter()
+            .map(|beta_tree| MerkleTreeRootDebug(beta_tree.root()))
+            .collect();
+
+        let alpha_tree_debug = MerkleTreeRootDebug(self.root);
+
+        f.debug_struct("HybridMerkleTree")
+            .field("beta_trees", &beta_trees_debug)
+            .field("alpha_tree", &alpha_tree_debug)
+            .field("root", &self.root)
+            .field("height", &self.height)
+            .field("alpha_tree_height", &self.alpha_tree_height)
+            .field("beta_tree_height", &self.beta_tree_height)
+            .field("n_leaves", &self.n_leaves)
+            .field("n_leaves_per_beta_tree", &self.n_leaves_per_beta_tree)
+            .field("n_beta_trees", &self.n_beta_trees)
+            .field("n_nodes", &self.n_nodes)
+            .field("last_node_index_per_layer", &self.last_node_index_per_layer)
+            .finish()
+    }
+}
+
+impl<AH, BH> PartialEq for HybridMerkleTree<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.root == other.root
+    }
+}
+
+impl<AH, BH> HybridMerkleTree<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    pub fn from_leaves<I>(leaves: I) -> Self
+    where
+        I: IntoIterator<Item = BH::Domain>,
+    {
+        let leaves: Vec<BH::Domain> = leaves.into_iter().collect();
+        let n_leaves = leaves.len();
+
+        if n_leaves < MIN_N_LEAVES {
+            panic!(
+                "Cannot create a Hybrid Merkle Tree with {} leaves",
+                n_leaves
+            );
+        }
+
+        if !is_pow2(n_leaves) {
+            panic!("Cannot create a Hybrid Merkle Tree with a non-power-of-two number of leaves")
+        }
+
+        let TreeInfo {
+            n_layers,
+            height,
+            alpha_tree_height,
+            beta_tree_height,
+            n_leaves,
+            n_leaves_per_beta_tree,
+            n_beta_trees,
+            n_nodes,
+            last_node_index_per_layer,
+        } = get_tree_info(n_leaves);
+
+        // Create the beta hasher's subtrees.
+        let beta_trees: Vec<MerkleTree<BH::Domain, BH::Function>> = leaves
+            .chunks(n_leaves_per_beta_tree)
+            .map(|beta_tree_leaves| {
+                let beta_tree_leaves = beta_tree_leaves.iter().cloned();
```

Comment:
Because `MerkleTree::new` takes an iterator of `T where T: Element`, not an iterator of `&T where T: Element` (the iterator used to create the Merkle Tree needs to have ownership of its items).